rpmlint.1
.TH RPMLINT "1" "April 2011" "rpmlint" "User Commands"
.SH NAME
rpmlint \- check common problems in rpm packages
.SH SYNOPSIS
\fBrpmlint\fR [\fIOPTION\fR]... [\fIFILE\fR|\fIPACKAGE\fR]...
.SH DESCRIPTION
\fBrpmlint\fR is a tool for checking common errors in rpm packages.
It can be used to test individual packages and spec files before
uploading or to check an entire distribution.  By default all
applicable checks are processed but specific checks can be performed
by using command line parameters.

\fIFILE\fR can be a rpm package file, a spec file, or a directory.  In
case of a directory, it is recursively searched for rpm and spec files
to check.  The special value \fB\-\fR results in standard input being
read and treated as (single) spec file content.
\fIPACKAGE\fR is the name of an installed package or a
.BR glob (7)
pattern to match installed packages, unless a file by that name exists.
.TP
\fB\-i\fR, \fB\-\-info\fR
Display explanations for reported messages.
.TP
\fB-I\fR, \fB\-\-explain\fR=\fImessageid\fR
Display explanations for the specified message identifiers and exit.
This option may be given multiple times.
.TP
\fB\-c\fR, \fB\-\-check\fR=\fIcheck\fR
Run only the specified check.  This option may be given multiple times
to specify multiple checks to run.  \fIcheck\fR is the name of the Python
module (as it would be given to Python's import statement) containing the
check.
.TP
\fB\-a\fR, \fB\-\-all\fR
Check all installed packages.
.TP
\fB\-C\fR, \fB\-\-checkdir\fR=\fIdir\fR
Insert \fIdir\fR to the front of the list of paths to load checks
from, unless it is already in the list.  The default list of check
dirs typically contains only /usr/share/rpmlint.  Directories in the
check dirs list are also inserted to the front of the list of paths to
load Python modules from when the check process begins.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display summary of command line options and exit.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Operate in verbose mode.
.TP
\fB\-E\fR, \fB\-\-extractdir\fR=\fIdir\fR
Base directory for extracted temporary files, default is what Python's
tempfile.gettempdir() returns.
.TP
\fB\-V\fR, \fB\-\-version\fR
Display version information and exit.
.TP
\fB\-n\fR, \fB\-\-noexception\fR
Ignore output filters.
.TP
\fB\-\-rawout\fR=\fIfile\fR
Write unfiltered output to \fIfile\fR.
.TP
\fB\-f\fR, \fB\-\-file\fR=\fIconffile\fR
Load user configuration from the specified file, default is
$XDG_CONFIG_HOME/rpmlint (~/.config/rpmlint if $XDG_CONFIG_HOME is
empty or not set).
.TP
\fB\-o\fR, \fB\-\-option\fR=\fIvalue\fR
Override a configuration option.  \fIvalue\fR is a whitespace separated string,
first word of which is the option name to set, and the Python eval() return
value for the rest is set as the value for the option.  Passing only an option
name is treated as if None was passed as its value.  See the file "config"
shipped with rpmlint for the list of configuration options and their types.
For example:
 \-o "NetworkEnabled True"
 \-o "Distribution \(aqMy favorite distro\(aq"
 \-o "MaxLineLength 80"
 \-o "ValidShells (\(aq/bin/sh\(aq, \(aq/bin/bash\(aq)"
.SH CAVEATS
All checks do not apply to all argument types.  For best check
coverage, run rpmlint on all source and binary packages your build
produces.  The set of checks rpmlint runs on source packages is a
superset of the one for plain specfiles, the set of checks run for
installed binary packages is a superset of the one for uninstalled
binary package files, and the source and binary package check sets are
quite different.
.SH FILES
.TP
\fB/usr/share/rpmlint/config\fR, \fB/usr/share/rpmlint/config.*\fR
Built-in configuration.  When invoked as \fIsomeprefix\fR-rpmlint,
/usr/share/rpmlint/config.\fIsomeprefix\fR is used if it exists,
otherwise /usr/share/rpmlint/config.
.TP
\fB/etc/rpmlint/*config\fR
System wide configuration.
.TP
\fB$XDG_CONFIG_HOME/rpmlint\fR or \fB~/.config/rpmlint\fR
User configuration.
.SH EXIT CODES
.IP 0
No errors.
.IP 1
Unspecified error.
.IP 2
Interrupted.
.IP 64
One or more error message printed.
.IP 66
Badness threshold exceeded.
.SH AUTHOR
Originally written by Frédéric Lepied, modified and maintained by
numerous contributors since.
.SH COPYRIGHT
This program is licensed under the GNU General Public License, see the
file COPYING included in the distribution archive.


rpmlint.py
# -*- coding: utf-8 -*-
#############################################################################
# File          : rpmlint.py
# Package       : rpmlint
# Author        : Frederic Lepied
# Created on    : Mon Sep 27 19:20:18 1999
# Version       : $Id$
# Purpose       : main entry point: process options, load the checks and run
#                 the checks.
#############################################################################

import getopt
import glob
import imp
import locale
import os
import re
import stat
import sys
import tempfile

# Do not import anything that initializes its global variables from
# Config at load time here (or anything that imports such a thing),
# that results in those variables initialized before config files are
# loaded which is too early - settings from config files won't take
# place for those variables.

from Filter import badnessScore, badnessThreshold, printAllReasons, \
     printDescriptions, printInfo, printed_messages
import AbstractCheck
import Config
import Pkg


_default_user_conf = '%s/rpmlint' % \
    (os.environ.get('XDG_CONFIG_HOME') or '~/.config')

# Print usage information
def usage(name):
    print ('''usage: %s [<options>] <rpm files|installed packages|specfiles|dirs>
  options:
\t[-i|--info]
\t[-I <messageid,messageid,...>]
\t[-c|--check <check>]
\t[-a|--all]
\t[-C|--checkdir <checkdir>]
\t[-h|--help]
\t[-v|--verbose]
\t[-E|--extractdir <dir>]
\t[-V|--version]
\t[-n|--noexception]
\t[-f|--file <user config file to use instead of %s]
\t[-o|--option <key value>]''' \
        % (name, _default_user_conf))

# Print version information
def printVersion():
    print ('rpmlint version %s Copyright (C) 1999-2007 Frederic Lepied, Mandriva' % Config.__version__)

def loadCheck(name):
    '''Load a (check) module by its name, unless it is already loaded.'''
    # Avoid loading more than once (initialization costs)
    loaded = sys.modules.get(name)
    if loaded:
        return loaded 
    (fobj, pathname, description) = imp.find_module(name)
    try:
        imp.load_module(name, fobj, pathname, description)
    finally:
        fobj.close()

#############################################################################
# main program
#############################################################################
def main():

    locale.setlocale(locale.LC_COLLATE, '')

    # Add check dirs to the front of load path
    sys.path[0:0] = Config.checkDirs()

    # Load all checks
    for c in Config.allChecks():
        loadCheck(c)

    packages_checked = 0
    specfiles_checked = 0
    do_spec_check = 'SpecCheck' in Config.allChecks()
    if do_spec_check:
        # See comments in "top level import section" for why this isn't
        # imported earlier.
        import SpecCheck

    try:
        # Loop over all file names given in arguments
        dirs = []
        for arg in args:
            pkgs = []
            isfile = False
            try:
                if arg == "-":
                    arg = "(standard input)"
                    # Short-circuit stdin spec file check
                    if do_spec_check:
                        stdin = sys.stdin.readlines()
                        if not stdin:
                            continue
                        pkg = Pkg.FakePkg(arg)
                        check = SpecCheck.SpecCheck()
                        check.verbose = verbose
                        check.check_spec(pkg, None, spec_lines=stdin)
                        pkg.cleanup()
                        specfiles_checked += 1
                    continue

                try:
                    st = os.stat(arg)
                    isfile = True
                    if stat.S_ISREG(st[stat.ST_MODE]):
                        if not arg.endswith(".spec"):
                            pkgs.append(Pkg.Pkg(arg, extract_dir))
                        elif do_spec_check:
                            # Short-circuit spec file checks
                            pkg = Pkg.FakePkg(arg)
                            check = SpecCheck.SpecCheck()
                            check.verbose = verbose
                            check.check_spec(pkg, arg)
                            pkg.cleanup()
                            specfiles_checked += 1

                    elif stat.S_ISDIR(st[stat.ST_MODE]):
                        dirs.append(arg)
                        continue
                    else:
                        raise OSError
                except OSError:
                    ipkgs = Pkg.getInstalledPkgs(arg)
                    if not ipkgs:
                        Pkg.warn(
                            '(none): E: no installed packages by name %s' % arg)
                    else:
                        ipkgs.sort(key = lambda x: locale.strxfrm(
                                x.header.sprintf("%{NAME}.%{ARCH}")))
                        pkgs.extend(ipkgs)
            except KeyboardInterrupt:
                if isfile:
                    arg = os.path.abspath(arg)
                Pkg.warn(
                    '(none): E: interrupted, exiting while reading %s' % arg)
                sys.exit(2)
            except Exception, e:
                if isfile:
                    arg = os.path.abspath(arg)
                Pkg.warn('(none): E: error while reading %s: %s' % (arg, e))
                pkgs = []
                continue

            for pkg in pkgs:
                runChecks(pkg)
                packages_checked += 1

        for dname in dirs:
            try:
                for path, dirs, files in os.walk(dname):
                    for fname in files:
                        fname = os.path.abspath(os.path.join(path, fname))
                        try:
                            if fname.endswith('.rpm') or \
                               fname.endswith('.spm'):
                                pkg = Pkg.Pkg(fname, extract_dir)
                                runChecks(pkg)
                                packages_checked += 1

                            elif do_spec_check and fname.endswith('.spec'):
                                pkg = Pkg.FakePkg(fname)
                                check = SpecCheck.SpecCheck()
                                check.verbose = verbose
                                check.check_spec(pkg, fname)
                                pkg.cleanup()
                                specfiles_checked += 1

                        except KeyboardInterrupt:
                            Pkg.warn('(none): E: interrupted, exiting while ' +
                                     'reading %s' % fname)
                            sys.exit(2)
                        except Exception, e:
                            Pkg.warn(
                                '(none): E: while reading %s: %s' % (fname, e))
                            continue
            except Exception, e:
                Pkg.warn(
                    '(none): E: error while reading dir %s: %s' % (dname, e))
                continue

        if printAllReasons():
            Pkg.warn('(none): E: badness %d exceeds threshold %d, aborting.' %
                     (badnessScore(), badnessThreshold()))
            sys.exit(66)

    finally:
        print "%d packages and %d specfiles checked; %d errors, %d warnings." \
              % (packages_checked, specfiles_checked,
                 printed_messages["E"], printed_messages["W"])

    if printed_messages["E"] > 0:
        sys.exit(64)
    sys.exit(0)

def runChecks(pkg):

    try:
        if verbose:
            printInfo(pkg, 'checking')

        for name in Config.allChecks():
            check = AbstractCheck.AbstractCheck.known_checks.get(name)
            if check:
                check.verbose = verbose
                check.check(pkg)
            else:
                Pkg.warn('(none): W: unknown check %s, skipping' % name)
    finally:
        pkg.cleanup()

#############################################################################
#
#############################################################################

sys.argv[0] = os.path.basename(sys.argv[0])

# parse options
try:
    (opt, args) = getopt.getopt(sys.argv[1:],
                              'iI:c:C:hVvanE:f:o:',
                              ['info',
                               'check=',
                               'checkdir=',
                               'help',
                               'version',
                               'verbose',
                               'all',
                               'noexception',
                               'extractdir=',
                               'file=',
                               'option=',
                               ])
except getopt.GetoptError, e:
    Pkg.warn("%s: %s" % (sys.argv[0], e))
    usage(sys.argv[0])
    sys.exit(1)

# process options
checkdir = '/usr/share/rpmlint'
checks = []
verbose = False
extract_dir = None
conf_file = _default_user_conf
if not os.path.exists(os.path.expanduser(conf_file)):
    # deprecated backwards compatibility with < 0.88
    conf_file = '~/.rpmlintrc'
info_error = None

# load global config files
configs = glob.glob('/etc/rpmlint/*config')
configs.sort()

# Was rpmlint invoked as a prefixed variant?
m = re.match(r"(?P<prefix>[\w-]+)-rpmlint(\.py)?", sys.argv[0])
if m:
    # Okay, we're a prefixed variant. Look for the variant config.
    # If we find it, use it. If not, fallback to the default.
    prefix = m.group('prefix')
    if os.path.isfile('/usr/share/rpmlint/config.%s' % prefix):
        configs.insert(0, '/usr/share/rpmlint/config.%s' % prefix)
    else:
        configs.insert(0, '/usr/share/rpmlint/config')
else:
    configs.insert(0, '/usr/share/rpmlint/config')

for f in configs:
    try:
        execfile(f)
    except IOError:
        pass
    except Exception, E:
        Pkg.warn('(none): W: error loading %s, skipping: %s' % (f, E))
# pychecker fix
del f

config_overrides = {}

# process command line options
for o in opt:
    if o[0] in ('-c', '--check'):
        checks.append(o[1])
    elif o[0] in('-i', '--info'):
        Config.info = True
    elif o[0] == '-I':
        info_error = o[1]
    elif o[0] in ('-h', '--help'):
        usage(sys.argv[0])
        sys.exit(0)
    elif o[0] in ('-C', '--checkdir'):
        Config.addCheckDir(o[1])
    elif o[0] in ('-v', '--verbose'):
        verbose = True
    elif o[0] in ('-V', '--version'):
        printVersion()
        sys.exit(0)
    elif o[0] in ('-E', '--extractdir'):
        extract_dir = o[1]
        Config.setOption('ExtractDir', extract_dir)
    elif o[0] in ('-n', '--noexception'):
        Config.no_exception = True
    elif o[0] in ('-a', '--all'):
        if '*' not in args:
            args.append('*')
    elif o[0] in ('-f', '--file'):
        conf_file = o[1]
    elif o[0] in ('-o', '--option'):
        kv = o[1].split(None, 1)
        if len(kv) == 1:
            config_overrides[kv[0]] = None
        else:
            config_overrides[kv[0]] = eval(kv[1])

# load user config file
try:
    execfile(os.path.expanduser(conf_file))
except IOError:
    pass
except Exception,E:
    Pkg.warn('(none): W: error loading %s, skipping: %s' % (conf_file, E))

# apply config overrides
for key, value in config_overrides.items():
    Config.setOption(key, value)

if not extract_dir:
    extract_dir = Config.getOption('ExtractDir', tempfile.gettempdir())

if info_error:
    Config.info = True
    sys.path[0:0] = Config.checkDirs()
    for c in checks:
        Config.addCheck(c)
    for c in Config.allChecks():
        loadCheck(c)
    for e in info_error.split(','):
        print "%s:" % e
        printDescriptions(e)
    sys.exit(0)

# if no argument print usage
if not args:
    usage(sys.argv[0])
    sys.exit(1)

if __name__ == '__main__':
    if checks:
        Config.resetChecks()
        for check in checks:
            Config.addCheck(check)
    main()

# rpmlint.py ends here

# Local variables:
# indent-tabs-mode: nil
# py-indent-offset: 4
# End:
# ex: ts=4 sw=4 et
